src/zenml/zen_server/rbac/utils.py

```diff
@@ -0,0 +1,569 @@
+        has_full_resource_access,
+        allowed_ids,
+    ) = rbac().list_allowed_resource_ids(
+        user=auth_context.user,
+        resource=Resource(type=resource_type),
+        action=action,
+    )
+
+    if has_full_resource_access:
+        return None
+
+    return {UUID(id) for id in allowed_ids}
+
+
+def get_resource_for_model(model: AnyResponseModel) -> Optional[Resource]:
+    """Get the resource associated with a model object.
+
+    Args:
+        model: The model for which to get the resource.
+
+    Returns:
+        The resource associated with the model, or `None` if the model
+        is not associated with any resource type.
+    """
+    resource_type = get_resource_type_for_model(model)
+    if not resource_type:
+        # This model is not tied to any RBAC resource type
+        return None
+
+    return Resource(type=resource_type, id=model.id)
+
+
+def get_surrogate_permission_model_for_model(
+    model: AnyResponseModel, action: str
+) -> Union[BaseResponse[Any, Any], BaseResponseModel]:
+    """Get a surrogate permission model for a model.
+
+    In some cases a different model instead of the original model is used to
+    verify permissions. For example, a parent container model might be used
+    to verify permissions for all its children.
+
+    Args:
+        model: The original model.
+        action: The action that the user wants to perform on the model.
+
+    Returns:
+        A surrogate model or the original.
+    """
+    from zenml.models import ModelVersionResponseModel
+
+    if action == Action.READ and isinstance(model, ModelVersionResponseModel):
+        # Permissions to read a model version is the same as reading the model
+        return model.model
+
+    return model
+
+
+def get_resource_type_for_model(
+    model: AnyResponseModel,
+) -> Optional[ResourceType]:
+    """Get the resource type associated with a model object.
+
+    Args:
+        model: The model for which to get the resource type.
+
+    Returns:
+        The resource type associated with the model, or `None` if the model
+        is not associated with any resource type.
+    """
+    from zenml.models import (
+        ArtifactResponse,
+        CodeRepositoryResponse,
+        ComponentResponse,
+        FlavorResponse,
+        ModelResponseModel,
+        PipelineResponse,
+        SecretResponseModel,
+        ServiceConnectorResponse,
+        StackResponse,
+    )
+
+    mapping: Dict[
+        Any,
+        ResourceType,
+    ] = {
+        FlavorResponse: ResourceType.FLAVOR,
+        ServiceConnectorResponse: ResourceType.SERVICE_CONNECTOR,
+        ComponentResponse: ResourceType.STACK_COMPONENT,
+        StackResponse: ResourceType.STACK,
+        PipelineResponse: ResourceType.PIPELINE,
+        CodeRepositoryResponse: ResourceType.CODE_REPOSITORY,
+        SecretResponseModel: ResourceType.SECRET,
+        ModelResponseModel: ResourceType.MODEL,
+        ArtifactResponse: ResourceType.ARTIFACT,
+    }
```

Comment:
I completely agree, even stuff like the "surrogate model" could be handled on the model classes themselves instead of having util functions that potentially contain too much knowledge.

The reason I did not implement it like this because I would have to duplicate everything on both the old and new base model classes until we convert the remaining models. If it's okay with you, I'd create a ticket to refactor this once we got rid of the old models.

Reply:
Fine by me, of course.